[WIP] Null when no associated records (fix sorting)

kbrock opened the pull request:
extracted:

- [x] #14 extracting tests - lots of the things floating around assume the tests are there.

for attributes with null values, doesn't go back to the database multiple times

for aggregates (currently only really use `sum`) - it returns 0

so when you sort by aggregates, the 0's (which are null behind the sceens) sort properly

this fixes an issue with #8

NickLaMuro commented on lib/active_record/virtual_attributes/virtual_total.rb:
```diff
@@ -72,21 +72,22 @@ def virtual_aggregate(name, relation, method_name = :sum, column = nil, options
+            else
+              rel = send(relation)
+              if rel.loaded?
+                rel.map { |t| t.send(column) }.compact.send(method_name)
```
Being blunt:  The commit message for this change is kind of a disaster... for reference:

> ### simplify loaded rel sum
> 
> array.map {} || 0 :
> 
> if there is an array, then it will always have a value,
>   so || 0 is never going to have a value
> 
> if an array is empty, empty.map {} will just be an empty value
> [].sum == 0
> 
> removed to_i since we could sum up floats

```diff
- rel.blank? ? nil : (rel.map { |t| t.send(column).to_i } || 0).send(method_name)
+ rel.map { |t| t.send(column) }.compact.send(method_name)
```


First off, the `compact` was completely unmentioned, but that plays a role in handling `nil` values that existed in the array previously (as `nil.to_i == 0`).

The `[] || 0` is actually a problem (if it was ever reachable) since `0.sum` is not a method on integers.

Finally, you removed `nil` being a valid return value, and that used to not be the case.

Regarding the formatting of the commit message, I think you would be better itemizing the changes (something like this, and probably could be tweaked):

```
simplify loaded rel sum

- removes `|| 0` since `0.sum` is invalid and it is not reachable (since `[] || 0 == []`)
- `[].sum == 0`, so we can avoid needing a `... || 0` case
- removes `nil` as being a possible return value when array is empty (now returns `0`)
- call `.compact` to remove `nil` values fromarray to avoid issue with `.send` (aka: `.sum`) later
```

* * *

The other issue I have is that we are now potentially returning floats when previously we casted to integers, and if that might potentially cause an issue down the line.

kbrock replied:
if the fields are string'd numbers, it worked before and does not work now. but not sure we should work around this. they screwed up.
if the fields were floats or bigints before, it would return a different value (or error) depending upon the source. Now it returns the correct value.

it still returns different values for unloaded vs loaded associations - 0 vs nil - this has not been addressed by this pr. Trying to keep this PR as a documentation/adding tests and less of a change.

kbrock replied:
I changed the comments. will split apart that commit

NickLaMuro commented on spec/virtual_total_spec.rb:
```diff
@@ -328,4 +345,45 @@ def model_with_children(count)
+      author
+      expect do
+        expect(author.sum_recently_published_books_rating).to eq(6)
+      end.to match_query_limit_of(1)
```
Can we have the test data for this test make this a `match_query_limit_of(2)` instead of `1`.

The  "calculates sum in database" has less of an impact when this spec calls the same number of queries.

kbrock replied:
@kbrock do you mean having 2 different authors? - so it shows the N+1?

NickLaMuro commented on spec/virtual_total_spec.rb:
```diff
@@ -328,4 +345,45 @@ def model_with_children(count)
+    it "calculates sum locally" do
+      author
+      expect do
+        expect(author.sum_recently_published_books_rating).to eq(6)
```
Aside:  This is a really dumb attribute... and I think I wrote it....

I feel that `avg_published_book_rating` would be something a little more realistic.

kbrock replied:
lol. was TOTALLY seeing the same thing.

Once we introduce average (which we don't in our app) - then the `nil` case really starts to show the issue of the `compact`

kbrock replied:
averages would be great - really show the issues with the nils

But then I'd have to come up with a sum test - and I don't want to deal with that right now :(
I'll make this change if you come up with another sum attribute

NickLaMuro commented:
@kbrock Knowing you, there is a important issue that I forgot to make more prominent in my review regarding the fact these totals now can be floats:

https://github.com/ManageIQ/activerecord-virtual_attributes/pull/13#discussion_r275924601

Please address that in the review response.

kbrock commented on lib/active_record/virtual_attributes/virtual_total.rb:
```diff
@@ -40,7 +40,7 @@ def virtual_total(name, relation, options = {})
       #    def allocated_disk_storage
       #      if disks.loaded?
-      #        disks.blank? ? nil : disks.map { |t| t.size.to_i }.sum
+      #        disks.map { |t| t.size }.compact.sum
```
thanks - I went ahead and removed that change.
it was a change, and this PR is more about trying to keep down those changes

kbrock replied:
removed this change - so this is all set

kbrock commented:
I fixed the codeclimate issue (even though I didn't break it on this code change)

I added a test to show the 0 vs nil for another edge case.
all (or most?) of the edge cases are now tested, so we can see the inconsistency in returning a 0 vs nil

The next step is to introduce a default value, so we can decide if we want a 0 or a nil when the relationships have no values. I like the default of a 0, but we do have cases that we seem to want a nil to come back.

I'll need to add quite a few tests around this, so adding defaults will take a little more time

kbrock commented on lib/active_record/virtual_attributes/virtual_total.rb:
```diff
@@ -72,21 +72,22 @@ def virtual_aggregate(name, relation, method_name = :sum, column = nil, options
-                  arel_column = rel.klass.arel_attribute(column)
-                  rel.try(method_name, arel_column) || 0
-                end
+            if has_attribute?(name)
```
good point - done

NickLaMuro commented on spec/virtual_total_spec.rb:
```diff
@@ -328,4 +361,65 @@ def model_with_children(count)
     end
   end
+
+  describe ".virtual_aggregation" do
```
Upon further inspection, not really, since they are both part of the same file.

Just seems like `virtual_total` inherits from `virtual_aggregation` conceptually, but instead, it is the reverse.

For now, carry on.

kbrock replied:
:( `virtual_total` calls `virtual_aggregate` - so yea, total should inherit from aggregation.

ideas how to make this more clear (probably a separate PR)

kbrock commented on lib/active_record/virtual_attributes/virtual_total.rb:
```diff
@@ -72,21 +72,21 @@ def virtual_aggregate(name, relation, method_name = :sum, column = nil, options
+              arel_column = rel.klass.arel_attribute(column)
+              rel.try(method_name, arel_column) || 0
+            elsif rel.blank?
+              nil
```
@NickLaMuro yes, change in direction - pairing this one back

NickLaMuro replied:
> One thing is for sure, returning different values based upon what what is loaded is a problem

So I think his needs to be addressed then before this is ready to go.

kbrock commented on spec/virtual_total_spec.rb:
```diff
@@ -265,6 +291,13 @@ def model_with_children(count)
       expect(model_with_children(2).v_total_named_books).to eq(2)
     end
 
+    it "falls back to default for busted virtual association" do
```
a virtual association that is not written correctly - busted.
e.g. it returns a nil instead of `[]` or `[a, b c]`

NickLaMuro replied:
My main question is "busted virtual association" is not very descriptive of what is being referenced, and there is no way for someone who isn't familiar with the code base (or your brain) to figure that out.

kbrock replied:
> falls back to default when virtual association is not written correctly in ruby

kbrock commented on lib/active_record/virtual_attributes/virtual_total.rb:
```diff
@@ -122,9 +122,13 @@ def virtual_aggregate_arel(reflection, method_name, column)
+          if method_name == :size
+            query
+          else
+            t.grouping(Arel::Nodes::NamedFunction.new('COALESCE', [query, Arel::Nodes::SqlLiteral.new("0")]))
```
thinking this should be part of a `default` value clause

NickLaMuro replied:
What do you mean?  As in this snippet:

```
Arel::Nodes::NamedFunction.new('COALESCE', [query, Arel::Nodes::SqlLiteral.new("0")])
```

Should moved to a helper method somewhere?  If so, I think that makes sense, though, arguably could wait for another PR if you don't want to tackle that now.

Up to you.

kbrock replied:
thinking we need to introduce a `:default => 0` for virtual attributes. if that was present, then we would use the coalesce to ensure the default value

kbrock commented:
Something to understand my reasoning behind

old: values.map { |v| v.to_i}.sum 
new: values.compact.sum

I see the second being more similar to the way that sql works.

```sql
create table numbers(num int);
insert into numbers (num) values (4), (8), (null);
select avg(num) from numbers; -- 6
select avg(num) from numbers where num is not null; -- 6
select avg(coalesce(num, 0)) from numbers; -- 4
```

`coalesce(num, 0)` is essentially the same as a sql `to_i` in our case.

method|`f(num)`|`f(num) !null`|`f(num.to_i)`
---|---|---|---
avg|6|6|4
min|4|4|0
sum|10|10|10
max|10|10|10

Note the `sum` case, the only case we use, ends up with the same value

kbrock commented:
closing in favor of #15
